### ISSUE TITLE
Project Abandonment and Revitalization

### ISSUE THREAD

**Issue by novacrazy** [brendanzab/color-rs#33]
Hey, if you aren't actively working on this I'd like to take over. I've rewritten color utilities at least five times now for separate projects, and I want to dedicate a crate to it once and for all.

If you require more details on what I plan to do with it, I can outline them all for you here.

**Comment by brendanzab** [comment]
Ooh, yes, would love to hear it!

**Comment by novacrazy** [comment]
Basically I want to provide what the [palette](https://github.com/Ogeon/palette) crate does not: an optimal pixel storage format that is suitable for textures and GPU interactions.

Something with a C array representation, but still usable with idiomatic Rust. Using arrays internally can allow for certain SIMD optimizations in release mode and with the target CPU instruction set specified.

Fortunately, the `nalgebra` crate taught be about a neat trick to dereference arrays as a component structure, and that would allow code like this:

```rust
let blue = RGB::new(0.0, 0.0, 1.0);
println!("{}", blue.b);
```

even though the color is stored internally as `[f32; 3]`. `nalgebra` uses this for vectors and matrices.

Furthermore, useful blend operations like OpenGL ones or common photoshop blend operations would be a big plus. [I've already implemented this once](https://github.com/combustion-engine/combustion/blob/master/combustion_common/src/color/blend.rs), so it's not a big deal to add here.

I would also like to provide a different approach to alpha channels and converting between alpha/non-alpha formats. My version would use associated types and type aliases to make `Alpha<RGB<T>>` equal to `RGBA<T>`, which is represented internally as `[T; 4]`, like so:

```rust
pub trait WithoutAlpha: Color {
    type WithAlpha: WithAlpha;

    fn with_alpha(self, alpha: Self::Channel) -> Self::WithAlpha;
}

pub trait WithAlpha: Color {
    type WithoutAlpha: WithoutAlpha;

    fn without_alpha(self) -> Self::WithoutAlpha;
}

pub type Alpha<T> = <T as WithoutAlpha>::WithAlpha;
```

then implement those for the correct color formats.

Channel types will stay pretty much the same, since we both just define them as typical scalars.

**Comment by brendanzab** [comment]
Nice thinking! Have you thought about how you would (or would not) handle colour spaces?

**Comment by novacrazy** [comment]
Hmm. To be honest I still haven't fully decided. RGB is really the only color space used on the GPU, and most blending operations I've used aren't well defined for HSV/HSL, let alone the others.

However, the palette crate supports HSL, HSV, HWB, Lab, LCh, Luma, RGB, XYZ and YxY. So perhaps a better solution would be to include convenience methods for converting between Palette types and this crate's type(s), placed behind a Cargo feature flag.

What do you think? Supporting only RGB isn't bad, since it's really all I or most programmers ever need with graphics programming.

Also, the projects that I intend to use this crate with are:

* whole game engine
* software renderer
* 3D model loader (replacement for Assimp)
* Custom texture format

and probably more. None of those use anything but RGB.

However, they do use sRGB, which is also something I'm unsure about handling.

If I were to use a wrapper type, it would have to perform all the operations of the normal linear colors, but do tons of conversion along the way, which would be slow.

Perhaps an alternative solution would be to place the conversion overhead on the user, and simply allow three methods for the `sRGB` wrapper type: `new`,  `from_linear` and `to_linear`, and immutable component access. That way users must convert back to linear to perform any operations, and no hidden conversions occur. In fact, the more I think about this the more I like it.

**Comment by novacrazy** [comment]
Don't get me wrong, if you would prefer we include other color spaces that's fine. However, since most of the work is already done over on the Palette crate, I'm indifferent to the extra work here.

**Comment by brendanzab** [comment]
Yeah definitely, I can appreciate you wanting to keep focused on your main domain. More thinking it might be good to leave room open to others to extend it in future contributions. PR #1 has more info if you are interested in past thoughts on the matter.

**Comment by brendanzab** [comment]
Have you thought of bringing some of your thoughts to the palette developers? Maybe there could be a possibility of joining forces with them? If not, I would be ok transferring ownership of this repo, and the crate to you. It's not likely I will get around to working on it in the near future, alas.

**Comment by brendanzab** [comment]
Hmm, interesting reading the ['What it isn't' section of the palette README](https://github.com/Ogeon/palette#what-it-isnt):

> # What It Isn't
>
> This library is only meant for color manipulation and conversion. It's not...
>
> * ...an image manipulation library. It will only handle colors, and not whole images.
> * ...an optimal pixel storage format. The colors are represented by linear floats, which is not a compact format and not meant to be displayed.
>
> You will have to look elsewhere for those particular features. There are,
> however, tools and traits in the [`pixel`][pixel_module] module for converting
> to and from various pixel formats. These are meant to work as bridges between
> Palette and other graphical libraries.

[pixel_module]: https://ogeon.github.io/docs/palette/master/palette/pixel/index.html

Would be interested to hear @Ogeon's thoughts on your ideas for color-rs!

**Comment by novacrazy** [comment]
I actually just opened up an issue over at Ogeon/palette/issues/71 for a new published version with all the stuff they've added in the last year. It's not a super heavy crate, so adding interoperability here with it would probably be our best bet for having additional color spaces. Palette's just not meant to be used for low level buffers and stuff. 

If anything, I might attempt to fork palette and rebuild it with the stuff I'm designing for this crate, like the array-based channel representations. It shouldn't add significant developer overhead with the whole `Deref` component magic.

**Comment by novacrazy** [comment]
For now, though, I'm going to fork color-rs and commit everything I'm working on for it so you can get a better idea of what I'm designing.

**Comment by Ogeon** [comment]
Hi! See that other issue for my, and by extension my projects', general status. Aside from that, I'm interested in any kind of collaboration to improve its use in game development and other GPU applications. My initial goal was just to make "correct color math" easy, so I decided to prioritize that over GPU interoperability. There are still some conversion interfaces to and from array formats, but not much more.

**Comment by novacrazy** [comment]
Hmm. Kind of an extreme suggestion, but if I am able to replace the core structures in Palette with the ones I'm working on here, would you be willing to consider merging color-rs and palette?

**Comment by Ogeon** [comment]
Your suggestion sounds pretty much lite palette with a more optimal storage solution. Maybe we can do some interface changes to allow things like in place conversion and ways to choose the internal representation. I think image does something similar. I have a hunch that it's possible to find a way.

**Comment by novacrazy** [comment]
I went ahead and put the "array" type I intent to use for channels into its own crate: [`numeric-array`](https://github.com/novacrazy/numeric-array). 

Totally didn't spend almost four days designing that and analyzing generated assembly code...

However, now I can define generic channels like this:

```rust
pub trait ChannelAssertion {
    type Channel: Channel;
}

impl<C> ChannelAssertion for C where C: Channel {
    type Channel = C;
}

pub type SingleChannel<C> = NumericArray<<C as ChannelAssertion>::Channel, U1>;
pub type DualChannel<C> = NumericArray<<C as ChannelAssertion>::Channel, U2>;
pub type TripleChannel<C> = NumericArray<<C as ChannelAssertion>::Channel, U3>;
pub type QuadChannel<C> = NumericArray<<C as ChannelAssertion>::Channel, U4>;
```

and it'll automatically make the best use of SIMD and so forth when doing operations on them.

Note that all operations on `NumericArray` are component-wise, so actual color blending is separate.

**Comment by rennis250** [comment]
I'm not sure I would remove support for transferring images in non-RGB formats to the GPU. I personally would benefit from running some algorithms on the GPU with colors in LAB or other spaces. I know that others in color vision research would also like this and why limit the possibilities of what one could do? I've already seen some people in the demo scene who have shader code for processing colors in LAB, LUV, and other spaces. I also know of at least one color vision researcher who uses non-RGB spaces on the GPU.

**Comment by brendanzab** [comment]
Oh, interesting. Thanks for mentioning that @rennis250!

I guess all a shader would want is a vector - it could convert it internally if needed. Of course its up to the shader to ensure that it interprets that vector correctly 🤔

**Comment by flying-sheep** [comment]
> RGB is really the only color space used on the GPU, and most blending operations I've used aren't well defined for HSV/HSL, let alone the others.

My use case is “have gradients that don’t look like ass”. Any color implementation that implements interpolation between two colors in sRGB space runs counter to that, since people will make the mistake to use it, resulting in gradients that look like ass.

So without color spaces, I support to firmly put interpolation as out of scope, and link a project that handles it instead. Interpolation without color spaces isn’t a well defined operation and will lead to naively generated gradients that look like ass.

**Comment by novacrazy** [comment]
It's been six years since I opened this issue. I've learned so much more about color since then. Color model != color space, all that jazz. You're correct that sRGB blending is crap.

It's also been 8 years since there was a commit here, so it's kind of moot. Everyone has moved to `palette` or custom solutions.
